frame/dapps-staking/src/tests.rs

```diff
@@ -1264,6 +1264,50 @@ fn new_era_forcing() {
     })
 }
 
+#[test]
+fn new_era_length_is_always_blocks_per_era() {
+    ExternalityBuilder::build().execute_with(|| {
+        initialize_first_block();
+        let blocks_per_era = mock::BLOCKS_PER_ERA;
+
+        // start new era era and record era number and height
+        advance_to_era(mock::DappsStaking::current_era() + 1);
+        let mut expected_era = mock::DappsStaking::current_era();
+        let mut previous_era_starting_block_height;
+        let mut current_era_starting_block_height = System::block_number();
+
+        // show normal era advancement works and its length is blocks_per_era
+        advance_to_era(mock::DappsStaking::current_era() + 1);
+        expected_era += 1;
+        previous_era_starting_block_height = current_era_starting_block_height;
+        current_era_starting_block_height = System::block_number();
+        assert_eq!(mock::DappsStaking::current_era(), expected_era);
+        assert_eq!(current_era_starting_block_height - previous_era_starting_block_height, blocks_per_era);
+
+
+        // now, move to middle of an era, force new era and assert the next block starts a new era
+        run_for_blocks(1);
+        <ForceEra<TestRuntime>>::put(Forcing::ForceNew);
+        run_for_blocks(1);  // era should increase here
+        expected_era += 1;
+        previous_era_starting_block_height = current_era_starting_block_height;
+        current_era_starting_block_height = System::block_number();
+        assert_eq!(mock::DappsStaking::current_era(), expected_era);
+
+        // show that the starting block of prev era and starting block of new era are less than blocks_per_era
+        assert!(current_era_starting_block_height - previous_era_starting_block_height < blocks_per_era);
+        assert!(current_era_starting_block_height - previous_era_starting_block_height > 0);
+        
+        // increment era and show the length of prev (forced) era is equal to blocks_per_era
+        advance_to_era(mock::DappsStaking::current_era() + 1);
+        expected_era += 1;
+        previous_era_starting_block_height = current_era_starting_block_height;
+        current_era_starting_block_height = System::block_number();
+        assert_eq!(mock::DappsStaking::current_era(), expected_era);
+        assert_eq!(current_era_starting_block_height - previous_era_starting_block_height, blocks_per_era);
```

Comment:
Two pieces of advice:
1) I'd suggest you break this UT into two separate tests: one for normal run and one that uses `force new era`. Makes it more clear what's being tested and easier to maintain in the future.
2) Reusing the same variable many times seems convenient but you risk a higher chance of making a mistake somewhere. In this particular scenario, you could end up modifying one variable (previous_era_starting_block) but forget about another (current_era_starting_block).

Reply:
1. Sure, I can split them into 2 tests.

2. I deviated from the "don't reuse a variable" rule because I thought it significantly increased readability (and therefore reliability) since pattern happens 3 times times in this test:

```rust
advance_to_era(mock::DappsStaking::current_era() + 1); // or something that should increases the era
expected_era += 1;
previous_era_starting_block_height = current_era_starting_block_height;
current_era_starting_block_height = System::block_number(); // update block height
assert_eq!(mock::DappsStaking::current_era(), expected_era); // make sure line 1 of code increased the current era
assert_eq!(current_era_starting_block_height - previous_era_starting_block_height, blocks_per_era); // make sure the distance between the blocks is what we expected
```

Readability was also same reason I left `previous_era_starting_block_height` uninitialized until line 1282. Makes the above pattern easier for the eyes to pick up on.

I originally was not reusing any variables, but readability became poorer as I had to find weird or super long variable names for storing and asserting the block numbers.